### PR TITLE
[NOT READY TO MERGE] Centralize the launch of models.run and add test on models.run

### PR DESCRIPTION
**Pull request by satyaog**
# Description

Centralize the launch of models.run and add test on models.run

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] `python run.py test`

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: run.py [hunk 1]
```diff
@@ -123,6 +123,14 @@ def tune(seed):
...
 
 
+@simu.command()
+def model():
+    import sys
+    from models.run import main as m_main
+    sys.argv = sys.argv[:1] + sys.argv[2:]
+    m_main()
+
+
```
Comment (satyaog) [inline run.py:133]: Not sure yet if it's good practice to plug models into the simu group

File: tests/models_test.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                        prev_unobserved = daily_output[current_day - 1][h_i][h_k]['unobserved']
+                    except IndexError:
+                        prev_observed = None
+                        prev_unobserved = None
+
+                    # Multi-hot arrays identifying the reported symptoms in the last 14 days
+                    # Symptoms:
+                    # ['aches', 'cough', 'fatigue', 'fever', 'gastro', 'loss_of_taste',
+                    #  'mild', 'moderate', 'runny_nose', 'severe', 'trouble_breathing']
+                    assert observed['reported_symptoms'].shape == (14, 11)
```
Comment (satyaog) [inline tests/models_test.py:69]: Currently, the shape is (14, 11+1). Why not 11?

File: tests/models_test.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                    # candidate_encounters[:, 0] is the other human 4 bits id
+                    # candidate_encounters[:, 1] is the risk of getting contaminated during the encounter?
+                    # candidate_encounters[:, 2] is the number of days since the encounter
+                    assert observed['candidate_encounters'].shape[1] == 3
+                    assert observed['candidate_encounters'][:, 0].min() >= 0
+                    assert observed['candidate_encounters'][:, 0].max() < 16
+                    assert observed['candidate_encounters'][:, 1].min() >= 0
+                    assert observed['candidate_encounters'][:, 1].max() < 16
+                    assert observed['candidate_encounters'][:, 2].min() >= 0
+                    assert observed['candidate_encounters'][:, 2].max() < 14
```
Comment (satyaog) [inline tests/models_test.py:79]: Currently `candidate_encounters[:, 2].max()` is <= 14. Why not < 14 if 0 based?

File: tests/models_test.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                    # Has received a positive test result [index] days before today
+                    assert unobserved['test_results'].shape == (14,)
+                    assert unobserved['test_results'].min() in (0, 1)
+                    assert unobserved['test_results'].max() in (0, 1)
+                    assert unobserved['test_results'].sum() in (0, 1)
```
Comment (satyaog) [inline tests/models_test.py:84]: `test_results.sum()` is probably `False` but just checking

File: tests/models_test.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+                    # Multi-hot arrays identifying the true symptoms in the last 14 days
+                    # Symptoms:
+                    # ['aches', 'cough', 'fatigue', 'fever', 'gastro', 'loss_of_taste',
+                    #  'mild', 'moderate', 'runny_nose', 'severe', 'trouble_breathing']
+                    assert unobserved['true_symptoms'].shape == (14, 11)
```
Comment (satyaog) [inline tests/models_test.py:90]: Currently, the shape is (14, 11+1). Why not 11?

File: tests/models_test.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+                    assert len(unobserved['exposed_locs'].shape) == 1
+                    assert unobserved['exposed_locs'].min() in (0, 1)
+                    assert unobserved['exposed_locs'].max() in (0, 1)
+                    assert 0 <= unobserved['exposed_locs'].sum() <= len(unobserved['exposed_locs'])
+                    # Encounters responsible for exposition. Exposition can occur without being
+                    # linked to an encounter
+                    assert len(unobserved['exposure_encounter'].shape) == 1
+                    assert unobserved['exposure_encounter'].min() in (0, 1)
+                    assert unobserved['exposure_encounter'].max() in (0, 1)
+                    assert unobserved['exposure_encounter'].sum() in (0, 1)
```
Comment (satyaog) [inline tests/models_test.py:113]: `exposure_encounter.sum()` is probably `False` but just checking

File: tests/models_test.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+                    # observed['reported_symptoms'] is a subset of unobserved['true_symptoms']
+                    assert (unobserved['true_symptoms'] == observed['reported_symptoms']) \
+                           [observed['reported_symptoms'].astype(np.bool)].all()
+
+                    if unobserved['is_infectious'] or unobserved['is_recovered']:
+                        assert unobserved['is_infectious'] != unobserved['is_recovered']
+
+                    # exposed_locs is the same length as candidate_locs
+                    # TODO: observed['candidate_locs'] should be a tuple (human_readable, id) preferably sorted
```
Comment (satyaog) [inline tests/models_test.py:127]: Let me know if this `TODO` should be implemented

File: tests/models_test.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+                    assert unobserved['exposed_locs'].shape == (len(observed['candidate_locs']),)
+
+                    # exposure_encounter is the same length as candidate_encounters
+                    assert unobserved['exposure_encounter'].shape == (observed['candidate_encounters'].shape[0],)
+
+                    if prev_observed:
+                        assert (observed['reported_symptoms'][:13, :] == prev_observed['reported_symptoms'][-13:, :]).all()
+                        assert (observed['candidate_encounters'][observed['candidate_encounters'][:, 2] > 1][:, 0:2] ==
```
Comment (satyaog) [inline tests/models_test.py:135]: this is currently `False` but I'm not sure why. It seams that `observed['candidate_encounters'][:, 2]` does not update with + 1 compared with `prev_observed['candidate_encounters'][:, 2]`

**Comment by satyaog** [comment]
I'm not entirely sure yet how I'm going to validate the results regarding the values in the logs.
Memory usage is too big for the test to run even with only 100 ppl and the test also times out. We might want to optimize it.

**Comment by mweiss17** [comment]
The functionality in this branch is now deprecated in favor of the new client-server architecture in the Interventions branch, which should be merged with develop by tonight. I believe almost all of this PR is contained in that branch. If we need the offline model prediction functionality in the next few days, we can use my master branch. After that, we should move towards always using risk prediction in the loop simulations.

**Pull request closed without merging**